File: riveredge-backend/src/apps/kuaizhizao/schemas/purchase_requisition.py
"""
采购申请数据验证Schema

Author: RiverEdge Team
Date: 2025-02-01
"""

from datetime import date, datetime
from typing import Dict, List, Optional
from decimal import Decimal

from pydantic import BaseModel, Field, ConfigDict, field_validator


class PurchaseRequisitionItemBase(BaseModel):
    """采购申请行基础"""
    model_config = ConfigDict(from_attributes=True)

    material_id: int = Field(..., description="物料ID")
    material_code: str = Field(..., max_length=50, description="物料编码")
    material_name: str = Field(..., max_length=200, description="物料名称")
    material_spec: Optional[str] = Field(None, max_length=200, description="物料规格")
    unit: str = Field("件", max_length=20, description="单位")
    quantity: Decimal = Field(..., gt=0, description="申请数量")
    suggested_unit_price: Decimal = Field(Decimal(0), ge=0, description="建议单价")
    required_date: Optional[date] = Field(None, description="要求到货日期")
    supplier_id: Optional[int] = Field(None, description="供应商ID")
    notes: Optional[str] = Field(None, description="备注")


class PurchaseRequisitionItemCreate(PurchaseRequisitionItemBase):
    """采购申请行创建"""
    demand_computation_item_id: Optional[int] = Field(None, description="需求计算明细ID")


class PurchaseRequisitionItemUpdate(BaseModel):
    """采购申请行更新"""
    quantity: Optional[Decimal] = Field(None, gt=0)
    suggested_unit_price: Optional[Decimal] = Field(None, ge=0)
    required_date: Optional[date] = None
    supplier_id: Optional[int] = None
    notes: Optional[str] = None


class PurchaseRequisitionItemResponse(PurchaseRequisitionItemBase):
    """采购申请行响应"""
    id: int
    requisition_id: int
    tenant_id: int
    demand_computation_item_id: Optional[int] = None
    purchase_order_id: Optional[int] = None
    purchase_order_item_id: Optional[int] = None
    supplier_id: Optional[int] = None
    converted_quantity_draft: Optional[Decimal] = Field(0, description="已下推数量（草稿）")
    converted_quantity_confirmed: Optional[Decimal] = Field(0, description="已下推数量（已确认）")
    created_at: datetime
    updated_at: datetime


class PurchaseRequisitionBase(BaseModel):
    """采购申请头基础"""
    model_config = ConfigDict(from_attributes=True)

    requisition_code: str = Field(..., max_length=50, description="申请编码")
    requisition_name: Optional[str] = Field(None, max_length=200, description="申请名称")
    status: str = Field("草稿", max_length=20, description="状态")
    applicant_id: Optional[int] = None
    applicant_name: Optional[str] = None
    requisition_date: Optional[date] = None
    required_date: Optional[date] = None
    source_type: Optional[str] = None
    source_id: Optional[int] = None
    source_code: Optional[str] = None
    is_urgent: bool = Field(False, description="是否紧急采购")
    urgent_reason: Optional[str] = None
    notes: Optional[str] = None


class PurchaseRequisitionCreate(BaseModel):
    """采购申请创建"""
    requisition_code: Optional[str] = None
    requisition_name: Optional[str] = None
    required_date: Optional[date] = None
    source_type: Optional[str] = None
    source_id: Optional[int] = None
    source_code: Optional[str] = None
    notes: Optional[str] = None
    items: List[PurchaseRequisitionItemCreate] = Field(..., description="申请明细")


class PurchaseRequisitionUpdate(BaseModel):
    """采购申请更新"""
    requisition_name: Optional[str] = None
    required_date: Optional[date] = None
    notes: Optional[str] = None
    items: Optional[List[PurchaseRequisitionItemCreate]] = None


class PurchaseRequisitionResponse(PurchaseRequisitionBase):
    """采购申请响应"""
    id: int
    tenant_id: int
    urgent_operator_id: Optional[int] = None
    urgent_operated_at: Optional[datetime] = None
    reviewer_id: Optional[int] = None
    reviewer_name: Optional[str] = None
    review_time: Optional[datetime] = None
    review_status: str = "待审核"
    review_remarks: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    items: List[PurchaseRequisitionItemResponse] = Field(default_factory=list)
    lifecycle: Optional[dict] = Field(None, description="生命周期（后端计算，供 UniLifecycleStepper 展示）")


class PurchaseRequisitionListResponse(PurchaseRequisitionResponse):
    """采购申请列表响应"""
    items_count: Optional[int] = None


class ConvertToPurchaseOrderRequest(BaseModel):
    """转采购单请求"""
    item_ids: List[int] = Field(..., description="要转单的采购申请行ID列表")
    supplier_id: Optional[int] = Field(None, description="统一供应商ID（行未指定且无 item_suppliers 映射时使用）")
    supplier_name: Optional[str] = Field(None, description="统一供应商名称（可选，缺省由服务端按 supplier_id 解析）")
    item_quantities: Optional[Dict[int, float]] = Field(None, description="按 item_id 覆盖数量，不传则用申请行原数量")
    item_suppliers: Optional[Dict[int, int]] = Field(
        None,
        description="申请行ID -> 供应商ID；优先于申请行上的 supplier_id 与顶层的 supplier_id，支持多供应商拆单",
    )
    persist_default_supplier_to_material: bool = Field(
        False,
        description="转单成功后，将本次下推使用的供应商写回采购件物料来源配置中的默认供应商",
    )

    @field_validator("item_suppliers", mode="before")
    @classmethod
    def _coerce_item_supplier_keys(cls, v):
        if v is None:
            return v
        if not isinstance(v, dict):
            return v
        return {int(k): int(val) for k, val in v.items()}


class ApproveRequisitionRequest(BaseModel):
    """审核采购申请请求"""
    approved: bool = Field(True, description="是否通过")
    review_remarks: Optional[str] = Field(None, description="审核备注/驳回原因")


class UrgentPurchaseRequest(BaseModel):
    """紧急采购请求"""
    urgent_reason: str = Field(..., min_length=1, description="紧急原因（必填）")

File: riveredge-frontend/src/apps/kuaizhizao/services/purchase-requisition.ts
/**
 * 采购申请API服务
 */

import { apiRequest } from '../../../services/api';

export interface ActionCapability {
  allowed: boolean;
  reason?: string | null;
}

export interface PurchaseRequisitionCapabilities {
  update?: ActionCapability;
  delete?: ActionCapability;
  submit?: ActionCapability;
  approve?: ActionCapability;
  revoke_approval?: ActionCapability;
}

export interface PurchaseRequisitionLifecycle {
  status_class?: string;
  flow_class?: string;
  current_stage_key?: string;
  current_stage_name?: string;
  status?: string;
  main_stages?: Array<{ key?: string; label?: string; status?: string }>;
  sub_stages?: Array<{ key?: string; label?: string; status?: string }>;
}

export interface PurchaseRequisitionItem {
  id?: number;
  material_id?: number;
  material_code?: string;
  material_name?: string;
  material_spec?: string;
  unit?: string;
  quantity?: number;
  suggested_unit_price?: number;
  required_date?: string;
  supplier_id?: number;
  purchase_order_id?: number;
  purchase_order_item_id?: number;
  demand_computation_item_id?: number;
  converted_quantity_draft?: number;
  converted_quantity_confirmed?: number;
  notes?: string;
}

export interface PurchaseRequisition {
  id?: number;
  requisition_code?: string;
  requisition_name?: string;
  status?: string;
  applicant_id?: number;
  applicant_name?: string;
  requisition_date?: string;
  required_date?: string;
  source_type?: string;
  source_id?: number;
  source_code?: string;
  review_status?: string;
  notes?: string;
  items_count?: number;
  items?: PurchaseRequisitionItem[];
  created_at?: string;
  updated_at?: string;
  lifecycle?: PurchaseRequisitionLifecycle;
  capabilities?: PurchaseRequisitionCapabilities;
}

export async function listPurchaseRequisitions(params: {
  skip?: number;
  limit?: number;
  status?: string;
  lifecycle_stage?: string;
  source_type?: string;
  keyword?: string;
  requisition_code?: string;
  requisition_name?: string;
  required_date_from?: string;
  required_date_to?: string;
} = {}): Promise<{ data: PurchaseRequisition[]; total: number; success: boolean }> {
  return apiRequest('/apps/kuaizhizao/purchase-requisitions', { method: 'GET', params });
}

export async function getPurchaseRequisition(id: number): Promise<PurchaseRequisition> {
  return apiRequest(`/apps/kuaizhizao/purchase-requisitions/${id}`);
}

export async function createPurchaseRequisition(data: Partial<PurchaseRequisition>): Promise<PurchaseRequisition> {
  return apiRequest('/apps/kuaizhizao/purchase-requisitions', { method: 'POST', data });
}

export async function updatePurchaseRequisition(id: number, data: Partial<PurchaseRequisition>): Promise<PurchaseRequisition> {
  return apiRequest(`/apps/kuaizhizao/purchase-requisitions/${id}`, { method: 'PUT', data });
}

export async function deletePurchaseRequisition(id: number): Promise<void> {
  return apiRequest(`/apps/kuaizhizao/purchase-requisitions/${id}`, { method: 'DELETE' });
}

export async function submitPurchaseRequisition(id: number): Promise<PurchaseRequisition> {
  return apiRequest(`/apps/kuaizhizao/purchase-requisitions/${id}/submit`, { method: 'POST' });
}

export async function approvePurchaseRequisition(
  id: number,
  data: { approved: boolean; review_remarks?: string }
): Promise<PurchaseRequisition> {
  return apiRequest(`/apps/kuaizhizao/purchase-requisitions/${id}/approve`, {
    method: 'POST',
    data,
  });
}

export async function withdrawPurchaseRequisition(id: number): Promise<PurchaseRequisition> {
  return apiRequest(`/apps/kuaizhizao/purchase-requisitions/${id}/withdraw-approval`, {
    method: 'POST',
  });
}

/** 撤回提交（待审核 -> 草稿），走统一审核入口 */
export async function withdrawPurchaseRequisitionSubmit(id: number): Promise<PurchaseRequisition> {
  return apiRequest(`/core/uni-audit/purchase_request/${id}/withdraw`, {
    method: 'POST',
  });
}

export async function fixPurchaseRequisitionStatus(id: number): Promise<PurchaseRequisition> {
  return apiRequest(`/apps/kuaizhizao/purchase-requisitions/${id}/fix-status`, {
    method: 'POST',
  });
}

export async function convertToPurchaseOrder(
  requisitionId: number,
  data: {
    item_ids: number[];
    supplier_id?: number;
    supplier_name?: string;
    item_quantities?: Record<number, number>;
    item_unit_prices?: Record<number, number>;
    /** 申请行 id -> 供应商 id，优先于行上 supplier_id */
    item_suppliers?: Record<number, number>;
    /** 转单成功后写回采购件物料默认供应商 */
    persist_default_supplier_to_material?: boolean;
  }
): Promise<{
  success: boolean;
  message: string;
  purchase_order_id: number;
  purchase_order_code: string;
  purchase_orders?: Array<{ purchase_order_id: number; purchase_order_code: string; supplier_id: number }>;
  persisted_material_ids?: number[];
}> {
  return apiRequest(`/apps/kuaizhizao/purchase-requisitions/${requisitionId}/convert-to-purchase-order`, {
    method: 'POST',
    data,
  });
}

export interface PriceComparisonItem {
  supplier_id: number;
  supplier_name: string;
  last_price: number;
  last_order_date?: string;
  delivery_lead_time: number;
}

export interface MaterialPriceComparison {
  material_id: number;
  material_code?: string;
  material_name: string;
  /** 后端字段名 comparison */
  comparison: PriceComparisonItem[];
}

export interface PriceComparisonResponse {
  results: MaterialPriceComparison[];
}

/** 获取多个物料的价格对比助手数据 */
export async function getPriceComparison(materialIds: number[]): Promise<PriceComparisonResponse> {
  return apiRequest('/apps/kuaizhizao/price-comparison', {
    method: 'GET',
    params: { material_ids: materialIds.join(',') },
  });
}
